src/imgui/ImGuiDebugger.cc

```diff
@@ -629,13 +629,39 @@ void ImGuiDebugger::drawDisassembly(CPURegs& regs, MSXCPUInterface& cpuInterface
 								}
 							});
 
+							enum Priority {
+								LOWEST_PRIORITY = 0,
+								MEDIUM_PRIORITY,
+								HIGHEST_PRIORITY
+							};
+							Priority currentPriority = LOWEST_PRIORITY;
+							std::vector<std::string_view> bestCandidates;
+							auto add_candidate = [&](const Symbol* sym, enum Priority newPriority) {
+								if (newPriority < currentPriority) return; // we already have a better candidate
+								if (newPriority > currentPriority) bestCandidates.clear(); // drop previous candidates, we found a better one
+								currentPriority = newPriority;
+								bestCandidates.push_back(sym->name); // cycle symbols in the same priority level
+							};
+
+							auto slot = getCurrentSlot(cpuInterface, debugger, addr16);
+							auto psss = (slot.ss.value_or(0) << 2) + slot.ps;
 							auto addrLabels = symbolManager.lookupValue(addr16);
-							std::string_view displayAddr = addrLabels.empty()
-								? std::string_view(addrStr)
-								: std::string_view(addrLabels[cycleLabelsCounter % addrLabels.size()]->name);
+							for (const Symbol* symbol: addrLabels) {
+								if (symbol->slot) {
+									if (*symbol->slot != psss) continue;
+									if (symbol->segment && symbol->segment != slot.seg) continue;
+									if (symbol->segment == slot.seg) {
+										add_candidate(symbol, HIGHEST_PRIORITY);
+									} else {
+										add_candidate(symbol, MEDIUM_PRIORITY);
+									}
+								} else if (!symbol->segment) {
```

Comment:
I was following what you said this earlier:

> I am assuming that symbols with segment information but without slot information don't make sense. We treat such symbols as if they don't have any segment info. Is this assumption correct?

>  if a symbol has no slot info, then it doesn't matter if it has segment info or not. We should always include it, no?. The current patch would never include symbols without slot but with segment info.

It does matter, otherwise any symbol from any slot will show. This is wrong behaviour because BLA should not be visible:
![image](https://github.com/user-attachments/assets/2cebc945-f8f9-4689-be1c-30f3ff046347)